contracts/Genesis.sol

```diff
@@ -82,107 +91,107 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
+     * @param count number of tokens to mint
+     * @param nonce nonce used to verify that the caller is allowed to mint
+     * @param proof Proof to verify that the caller is allowed to mint
      */
-    // TODO Update function to properly use the requests
-    // TODO: Add logic for the reserved spots
-    function mintNFTs(uint256 _count) public payable {
-        require(mintActive, "Minting is not active yet!");
-        uint256 mintIndex = _nextTokenId.current();
-        require(mintIndex + _count <= MAX_SUPPLY, "NFTs sold out");
-        require(msg.value >= PRICE * _count, "Not enough ETH");
-
-        for (uint256 i = 0; i < _count; i++) {
-            // TODO Validate this
-            mintIndex = _nextTokenId.current();
-            _nextTokenId.increment();
-            requestRandomNumberForTokenId(mintIndex);
-        }
+    function freeMint(uint256 count, uint256 nonce, bytes32[] calldata proof) public payable {
+        // TODO
     }
 
-    // TODO: Add logic for the reserved spots
-    function mint() public payable returns (bytes32) {
+    /**
+     * Mint 1 token
+     * The sender has to be included in the whitelist (except if they are the owner)
+     * @param nonce nonce used to verify that the caller is allowed to mint
+     * @param proof Proof to verify that the caller is allowed to mint
+     * @return requestId request id that was sent to Chainlink VRF
+     */
+    function mintWhitelist(uint256 nonce, bytes32[] calldata proof) public payable onlyWhitelist(nonce, proof) returns (bytes32 requestId) {
         require(mintActive, "Minting is not active yet!");
         uint256 mintIndex = _nextTokenId.current();
-        require(mintIndex <= MAX_SUPPLY, "NFTs sold out");
+        require(mintIndex <= MAX_SUPPLY, "Sold out");
         require(msg.value >= PRICE, "Not enough ETH");
-        require(
-            addressToMintCount[msg.sender] < MAX_PER_MINT,
-            "No more minting spot left"
-        );
-
+        require(tokenMintedCount[msg.sender] == 0, "Already minted");
+        tokenMintedCount[msg.sender]++;
         _nextTokenId.increment();
-        bytes32 requestId = requestRandomNumberForTokenId(mintIndex);
-        addressToMintCount[msg.sender] = addressToMintCount[msg.sender].add(1);
+        requestId = requestRandomNumberForTokenId(mintIndex);
         emit RequestedRandomNFT(requestId);
         return requestId;
     }
 
-    function requestRandomNumberForTokenId(uint256 _tokenId)
-        internal
-        returns (bytes32 requestId)
-    {
-        require(
-            LINK.balanceOf(address(this)) >= fee,
-            "Not enough LINK - fill contract with faucet"
-        );
+    /**
+     * Requests a random number to Chainlink VRF for a given token id
+     * @param tokenId id of the token
+     * @return requestId id of the request sent to Chainlink
+     */
+    function requestRandomNumberForTokenId(uint256 tokenId) internal returns (bytes32 requestId) {
```

Comment:
Just noticed here but fuck the convention of using `_` before function param? If so, maybe change the `constructor`? You never used it anywhere, it's fine with me, but we should be consistent.

Reply:
updated tall occurences